File: clients/kmp/shared/src/commonMain/kotlin/org/wyrdsekai/app/engine/room/RoomState.kt
package org.wyrdsekai.app.engine.room

import org.wyrdsekai.app.engine.event.WorldEvent
import org.wyrdsekai.app.protocol.*

/**
 * Mutable room state derived from event stream via apply().
 * Port of the event-folding logic in core/room/RoomActor.java.
 */
data class RoomState(
    val roomId: String,
    val name: String,
    val description: String,
    val zone: String,
    val exits: Map<String, Exit>,
    val entities: Map<String, Entity>,
    val objects: Map<String, RoomObject>,
    val hints: List<Hint>,
    val properties: Map<String, String>,
) {
    companion object {
        fun empty(roomId: String) = RoomState(
            roomId = roomId,
            name = "",
            description = "",
            zone = "",
            exits = emptyMap(),
            entities = emptyMap(),
            objects = emptyMap(),
            hints = emptyList(),
            properties = emptyMap(),
        )
    }

    /** Pure function: apply an event to produce new state. */
    fun apply(event: WorldEvent): RoomState = when (event) {
        is WorldEvent.RoomCreated -> copy(
            name = event.name,
            description = event.description,
            zone = event.zone,
        )
        is WorldEvent.EntityEntered -> copy(
            entities = entities + (event.entityId to Entity(
                id = event.entityId,
                name = event.entityName,
                type = event.entityType,
                description = "",
            ))
        )
        is WorldEvent.EntityLeft -> copy(entities = entities - event.entityId)
        is WorldEvent.ObjectTaken -> copy(objects = objects - event.objectId)
        is WorldEvent.ObjectDropped -> copy(
            objects = objects + (event.objectId to RoomObject(
                id = event.objectId,
                name = event.objectName,
                description = event.description,
                takeable = event.takeable,
            ))
        )
        is WorldEvent.ObjectAdded -> copy(
            objects = objects + (event.objectId to RoomObject(
                id = event.objectId,
                name = event.objectName,
                description = event.description,
                takeable = event.takeable,
            ))
        )
        is WorldEvent.ExitOpened -> copy(
            exits = exits + (event.direction to Exit(
                direction = event.direction,
                targetRoom = event.targetRoom,
                label = event.label,
            ))
        )
        is WorldEvent.ExitClosed -> copy(exits = exits - event.direction)
        is WorldEvent.DescriptionChanged -> copy(description = event.newDescription)
        is WorldEvent.HintsUpdated -> copy(hints = event.hints)
        is WorldEvent.PropertyChanged -> copy(
            properties = if (event.newValue == null) {
                properties - event.key
            } else {
                properties + (event.key to event.newValue)
            }
        )
        // Events that don't change room state
        is WorldEvent.Said,
        is WorldEvent.Emoted,
        is WorldEvent.ObjectUsed,
        is WorldEvent.ScriptTriggered,
        is WorldEvent.Whispered -> this
    }

    /** Convert to protocol snapshot for wire transmission. */
    fun toSnapshot(): RoomSnapshot = RoomSnapshot(
        roomId = roomId,
        name = name,
        description = description,
        zone = zone,
        exits = exits.values.toList(),
        entities = entities.values.toList(),
        objects = objects.values.toList(),
        hints = hints,
    )
}

File: clients/rn/src/engine/room/RoomState.ts
/**
 * Mutable room state derived from event stream via apply().
 * TypeScript port of KMP's RoomState.kt.
 */

import type { Exit, Entity, RoomObject, Hint, RoomSnapshot } from '../../protocol/models';
import type { WorldEvent } from '../events/WorldEvent';

export interface RoomState {
  roomId: string;
  name: string;
  description: string;
  zone: string;
  exits: Record<string, Exit>;
  entities: Record<string, Entity>;
  objects: Record<string, RoomObject>;
  hints: Hint[];
  properties: Record<string, string>;
}

export function emptyRoomState(roomId: string): RoomState {
  return {
    roomId,
    name: '',
    description: '',
    zone: '',
    exits: {},
    entities: {},
    objects: {},
    hints: [],
    properties: {},
  };
}

/** Pure function: apply an event to produce new state. */
export function applyEvent(state: RoomState, event: WorldEvent): RoomState {
  switch (event.type) {
    case 'room_created':
      return { ...state, name: event.name, description: event.description, zone: event.zone };

    case 'entity_entered':
      return {
        ...state,
        entities: {
          ...state.entities,
          [event.entityId]: {
            id: event.entityId,
            name: event.entityName,
            type: event.entityType as Entity['type'],
            description: '',
          },
        },
      };

    case 'entity_left': {
      const { [event.entityId]: _, ...rest } = state.entities;
      return { ...state, entities: rest };
    }

    case 'object_taken': {
      const { [event.objectId]: _, ...rest } = state.objects;
      return { ...state, objects: rest };
    }

    case 'object_dropped':
      return {
        ...state,
        objects: {
          ...state.objects,
          [event.objectId]: {
            id: event.objectId,
            name: event.objectName,
            description: event.description,
            takeable: event.takeable,
          },
        },
      };

    case 'object_added':
      return {
        ...state,
        objects: {
          ...state.objects,
          [event.objectId]: {
            id: event.objectId,
            name: event.objectName,
            description: event.description,
            takeable: event.takeable,
          },
        },
      };

    case 'exit_opened':
      return {
        ...state,
        exits: {
          ...state.exits,
          [event.direction]: {
            direction: event.direction,
            targetRoom: event.targetRoom,
            label: event.label,
          },
        },
      };

    case 'exit_closed': {
      const { [event.direction]: _, ...rest } = state.exits;
      return { ...state, exits: rest };
    }

    case 'description_changed':
      return { ...state, description: event.newDescription };

    case 'hints_updated':
      return { ...state, hints: event.hints };

    case 'property_changed': {
      if (event.newValue == null) {
        const { [event.key]: _, ...rest } = state.properties;
        return { ...state, properties: rest };
      }
      return { ...state, properties: { ...state.properties, [event.key]: event.newValue } };
    }

    // Events that don't change room state
    case 'said':
    case 'emoted':
    case 'object_used':
    case 'script_triggered':
    case 'whispered':
    case 'vitality_suggested':
      return state;
  }
}

/** Convert to protocol snapshot for wire transmission. */
export function toSnapshot(state: RoomState): RoomSnapshot {
  return {
    roomId: state.roomId,
    name: state.name,
    description: state.description,
    zone: state.zone,
    exits: Object.values(state.exits),
    entities: Object.values(state.entities),
    objects: Object.values(state.objects),
    hints: state.hints,
  };
}
